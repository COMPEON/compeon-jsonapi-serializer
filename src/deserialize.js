import { each, find, get, isPlainObject } from 'lodash'

const getId = (id, lid) => {
  if (id) return { id: String(id) }
  return { lid: String(lid) }
}

export const deserialize = (options = {}) => object => {
  const included = get(object, 'included', {})
  const data = get(object, 'data')

  if (Array.isArray(data)) {
    const results = []

    each(data, resource => {
      const id = get(resource, 'id')
      const result = get(resource, 'attributes', {})
      const relationships = get(resource, 'relationships', {})

      each(relationships, (relationship, relationshipName) => {
        const data = get(relationship, 'data')

        if (Array.isArray(data)) {
          const finalRelationships = []
          each(data, (relationshipDataSet, index) => {
            const { id, lid, type } = relationshipDataSet

            const include = find(included, {
              ...getId(id, lid),
              type
            })

            const dissolvedRelationship = include
              ? { ...getId(id, lid), ...include.attributes }
              : getId(id, lid)

            finalRelationships.push(dissolvedRelationship)
          })

          result[relationshipName] = finalRelationships
        } else if (isPlainObject(data)) {
          const { id, lid, type } = data

          const include = find(included, {
            ...getId(id, lid),
            type
          })

          const dissolvedRelationship = include
            ? { ...getId(id, lid), ...include.attributes }
            : getId(id, lid)

          result[relationshipName] = dissolvedRelationship
        }
      })

      if (id !== undefined) result.id = id

      results.push(result)
    })

    return results
  } else {
    const id = get(data, 'id')
    const result = get(data, 'attributes', {})
    const relationships = get(data, 'relationships', {})

    each(relationships, (relationship, relationshipName) => {
      const data = get(relationship, 'data')

      if (Array.isArray(data)) {
        const finalRelationships = []
        each(data, (relationshipDataSet, index) => {
          const { id, lid, type } = relationshipDataSet

          const include = find(included, {
            ...getId(id, lid),
            type
          })

          const dissolvedRelationship = include
            ? { ...getId(id, lid), ...include.attributes }
            : getId(id, lid)

          finalRelationships.push(dissolvedRelationship)
        })

        result[relationshipName] = finalRelationships
      } else if (isPlainObject(data)) {
        const { id, lid, type } = data

        const include = find(included, {
          ...getId(id, lid),
          type
        })

        const dissolvedRelationship = include
          ? { ...getId(id, lid), ...include.attributes }
          : getId(id, lid)

        result[relationshipName] = dissolvedRelationship
      }
    })

    if (id !== undefined) result.id = id

    return result
  }
}
