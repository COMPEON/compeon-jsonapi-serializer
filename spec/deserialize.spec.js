import { deserialize } from 'deserialize'

describe('deserialize', () => {
  describe('with simple attributes', () => {
    describe('with valid attributes', () => {
      const json = {
        data: {
          attributes: {
            firstName: 'Nico',
            lastName: 'Peters'
          },
          id: '123',
          type: 'users'
        }
      }

      it('deserializes the json', () => {
        expect(deserialize()(json)).toMatchSnapshot()
      })
    })

    describe('with invalid attributes', () => {
      const json = {
        data: {}
      }

      it('deserializes the json', () => {
        expect(deserialize()(json)).toMatchSnapshot()
      })
    })
  })
})
