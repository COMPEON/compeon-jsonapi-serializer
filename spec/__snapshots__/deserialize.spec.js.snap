// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deserialize with a 1:1 relationship with a local id deserializes the json 1`] = `
Object {
  "company": Object {
    "lid": "666",
  },
  "firstName": "Nico",
  "id": "123",
  "lastName": "Peters",
}
`;

exports[`deserialize with a 1:1 relationship with a normal id deserializes the json 1`] = `
Object {
  "company": Object {
    "id": "666",
  },
  "firstName": "Nico",
  "id": "123",
  "lastName": "Peters",
}
`;

exports[`deserialize with a 1:1 relationship with includes with a local id deserializes the json 1`] = `
Object {
  "company": Object {
    "city": "Düsseldorf",
    "lid": "666",
    "name": "Compeon GmbH",
  },
  "firstName": "Nico",
  "id": "123",
  "lastName": "Peters",
}
`;

exports[`deserialize with a 1:1 relationship with includes with a normal id deserializes the json 1`] = `
Object {
  "company": Object {
    "city": "Düsseldorf",
    "id": "666",
    "name": "Compeon GmbH",
  },
  "firstName": "Nico",
  "id": "123",
  "lastName": "Peters",
}
`;

exports[`deserialize with a 1:n relationship with a local id deserializes the json 1`] = `
Object {
  "companies": Array [
    Object {
      "lid": "666",
    },
    Object {
      "id": "667",
    },
  ],
  "firstName": "Nico",
  "id": "123",
  "lastName": "Peters",
}
`;

exports[`deserialize with a 1:n relationship with a normal id deserializes the json 1`] = `
Object {
  "companies": Array [
    Object {
      "id": "666",
    },
    Object {
      "id": "667",
    },
  ],
  "firstName": "Nico",
  "id": "123",
  "lastName": "Peters",
}
`;

exports[`deserialize with a 1:n relationship with includes with a local id deserializes the json 1`] = `
Object {
  "companies": Array [
    Object {
      "city": "Düsseldorf",
      "lid": "666",
      "name": "Compeon GmbH",
    },
    Object {
      "city": "Düsseldorf",
      "id": "667",
      "name": "Compeon 4.0 GmbH",
    },
  ],
  "firstName": "Nico",
  "id": "123",
  "lastName": "Peters",
}
`;

exports[`deserialize with a 1:n relationship with includes with a normal id deserializes the json 1`] = `
Object {
  "companies": Array [
    Object {
      "city": "Düsseldorf",
      "id": "666",
      "name": "Compeon GmbH",
    },
    Object {
      "city": "Düsseldorf",
      "id": "667",
      "name": "Compeon 4.0 GmbH",
    },
  ],
  "firstName": "Nico",
  "id": "123",
  "lastName": "Peters",
}
`;

exports[`deserialize with an array of resources with valid attributes deserializes the json 1`] = `
Array [
  Object {
    "firstName": "Nico",
    "id": "123",
    "lastName": "Peters",
  },
  Object {
    "firstName": "Frank",
    "id": "234",
    "lastName": "Wüller",
  },
]
`;

exports[`deserialize with empty relationships ignores the relationship 1`] = `
Object {
  "firstName": "Nico",
  "id": "123",
  "lastName": "Peters",
}
`;

exports[`deserialize with relationships inside an array of resources deserializes the json 1`] = `
Array [
  Object {
    "company": Object {
      "city": "Düsseldorf",
      "lid": "666",
      "name": "Compeon GmbH",
    },
    "firstName": "Nico",
    "id": "123",
    "lastName": "Peters",
  },
  Object {
    "company": Object {
      "city": "Düsseldorf",
      "lid": "666",
      "name": "Compeon GmbH",
    },
    "firstName": "Frank",
    "id": "234",
    "lastName": "Wüller",
    "tags": Array [
      Object {
        "id": "912",
        "name": "Banking",
      },
      Object {
        "id": "56",
        "name": "CEO",
      },
    ],
  },
]
`;

exports[`deserialize with simple attributes with invalid attributes when data does not exist deserializes the json 1`] = `Object {}`;

exports[`deserialize with simple attributes with invalid attributes when data is empty deserializes the json 1`] = `Object {}`;

exports[`deserialize with simple attributes with valid attributes deserializes the json 1`] = `
Object {
  "firstName": "Nico",
  "id": "123",
  "lastName": "Peters",
}
`;
